clouddq/templates/dbt/macros/macros.sql

```diff
@@ -53,7 +55,91 @@
     zero_record.rule_binding_id = data.rule_binding_id
 {% endmacro -%}
 
-{% macro validate_complex_rule(rule_id, rule_configs, rule_binding_id, column_name, fully_qualified_table_name, include_reference_columns ) -%}
+{% macro validate_complex_rule(rule_id, rule_configs, rule_binding_id, column_name, fully_qualified_table_name, include_reference_columns, configs) -%}
+  SELECT
+    CURRENT_TIMESTAMP() AS execution_ts,
+    '{{ rule_binding_id }}' AS rule_binding_id,
+    '{{ rule_id }}' AS rule_id,
+    '{{ fully_qualified_table_name }}' AS table_id,
+    CAST(NULL AS STRING) AS column_id,
+    NULL AS column_value,
+    {% for ref_column_name in include_reference_columns %}
+        custom_sql_statement_validation_errors.{{ ref_column_name }} AS {{ ref_column_name }},
+    {%- endfor -%}
+{% if rule_configs.get("dimension") %}
+    '{{ rule_configs.get("dimension") }}' AS dimension,
+{% else %}
+    CAST(NULL AS STRING) AS dimension,
+{% endif %}
+    CAST(NULL AS BOOLEAN) AS simple_rule_row_is_valid,
+    TRUE AS skip_null_count,
+    custom_sql_statement_validation_errors.complex_rule_validation_errors_count AS complex_rule_validation_errors_count,
+    CASE
+      WHEN custom_sql_statement_validation_errors.complex_rule_validation_errors_count IS NULL THEN CAST(NULL AS BOOLEAN)
+      WHEN custom_sql_statement_validation_errors.complex_rule_validation_errors_count = 0 THEN TRUE
+      ELSE FALSE
+    END AS complex_rule_validation_success_flag,
+    """{{ configs.get(rule_binding_id ~ '_' ~ rule_id ~ '_failed_records_sql_string') }}"""
+    AS failed_records_query,
+  FROM
+    zero_record
+  LEFT JOIN
+    (
+      SELECT
+         *,
+        '{{ rule_binding_id }}' AS _rule_binding_id,
+        COUNT(*) OVER() AS complex_rule_validation_errors_count,
+      FROM (
+      {{ rule_configs.get("rule_sql_expr") }}
+      ) custom_sql
+    ) custom_sql_statement_validation_errors
+  ON
+    zero_record.rule_binding_id = custom_sql_statement_validation_errors._rule_binding_id
+{% endmacro -%}
+
+
+{% macro validate_simple_rule_failed_records_query(rule_id, rule_configs, rule_binding_id, column_name, fully_qualified_table_name, include_reference_columns) -%}
```

Comment:
So..validate_simple_rule_failed_records_query and validate_complex_rule_failed_records_query do not have "failed_records_query" as part of SELECT, while validate_simple_rule and validate_complex_rule do?

When should "validate_simple_rule_failed_records_query" be used? And when "validate_simple_rule" should be used?

Reply:
The "**validate_simple_rule_failed_records_query**" is used in **failed_records_query** template and the "**validate_simple_rule**" will be used in **rule_binding_template**. The failed_records_query template should not have "failed_records_query" field. This template will be added as a field in rule_binding_per_rule.